src/twisted/internet/defer.py

```diff
@@ -437,6 +437,12 @@ class Deferred(Awaitable[_SelfResultT]):
     # used.
     _runningCallbacks = False
 
+    # Used to optimize isinstance(obj, Deferred) by
```

Comment:
I think that is best to document this as an `@ivar` as part of the class docstring.

Similar to `@ivar _suppressAlreadyCalled:` 

In this way, this information will be part of our HTML API docs ... it should be a bit more visbile

Reply:
Doesn't this entice users to rely on this mechanism in their code? They don't know whether `_` prefix means that the API is _really_ private. If a variable or function is documented this suggest that it's part of the API.

Reply:
It's not really private API according to the documented or prevailing conventions.  The `_` prefix makes the `_twistedPrivateIsDeferred` attribute private to `Deferred` which means applications shouldn't touch it and expect Twisted to continue supporting it according to the backwards compatibility policy.  But the new implementation looks at this attribute on *any object that is returned by a callback*.  Those objects are free to have their own attributes, private or otherwise.  Twisted's compatibility policy doesn't say anything about what attributes applications have on their own objects.

If we want this new behavior to really be an implementation detail, not something applications depend on, we should probably document it separately and call it out as an optimisation, not a guaranteed semantic.  This probably means putting it in one of the Deferred howtos or the Deferred class docstring.

Personally, I wonder if the performance benefit is worth the extra complexity.  I also wonder how much of the speedup comes from the true path through this code and how much comes from the false path.  The false path can be sped up with no risk of semantic changes by _keeping_ the existing test but checking for this magic attribute first.  If the attribute isn't present, the existing test will be skipped and some speedup should result.  If the attribute is present then the object still must be a real Deferred or the next check will fail.  This means there's no speedup for the Deferred case but there's also no chance an application shoves a non-Deferred in here that fools the code into thinking it is a Deferred.

I also wonder if a real port of Deferred to, eg, Rust, would be a way to keep the desired semantics, a simple implementation, and get the maximum possible performance.

Reply:
Apropos of my other recent comment, I feel like it would be mostly okay to claim the namespace `__twisted_private_.*__` for other objects, if we needed such a thing.  Like if somebody went out of their way to name an attribute `__twisted_private_isDeferred__` then they deserve what's coming to them much more than e.g. `__table_args__` or `__providedBy__` or `__provides__`, which are other attributes that other libraries will staple to your objects or classes without telling you about them or where they're from.

Reply:
> Personally, I wonder if the performance benefit is worth the extra complexity. I also wonder how much of the speedup comes from the true path through this code and how much comes from the false path. The false path can be sped up with no risk of semantic changes by keeping the existing test but checking for this magic attribute first. 

Both false and true paths are important. True path is used in inlineCallbacks when it yields another Deferred (e.g. coming from another inlineCallbacks decorated function). It is very frequent use case.

Reply:
> Both false and true paths are important. True path is used in inlineCallbacks when it yields another Deferred (e.g. coming from another inlineCallbacks decorated function). It is very frequent use case.

Would you mind doing the experiment (make the change and push the result, see what codspeed thinks) just so we can put a number on it?

If we do this and codspeed thinks it's the same, then that suggests we need better benchmarks